lazyClient.go

```diff
@@ -147,6 +147,22 @@ func (l *lazyClientConn[T]) Close() error {
 	// closed the stream for reading. I mean, we're the initiator so that's
 	// strange... but it's still allowed
 	_ = l.Flush()
+
+	// Finish reading the handshake before we close the connection/stream. This
+	// is necessary so that the other side can finish sending its response to our
+	// multistream header before we tell it we are done reading.
+	//
+	// Example:
+	// We open a QUIC stream, write the protocol `/a`, send 1 byte of application
+	// data, and immediately close.
```

Comment:
Do we hit this case when, for example, we open a stream and then reset it immediately due to an application error?

Reply:
It would. But in that case, it's fine to not read the handshake message, as the other side may or may not receive the writes made before calling `Reset`.

Reply:
I was confused when this case happened in practice in our protocol, but I realized that Bitswap opens a streams, writes to it and closes it, potentially never reading handshake data.

Reply:
For completeness, clients can handle this situation with a fix like: https://github.com/multiformats/go-multistream/pull/87 which ignores a reset on writing the multistream header, allowing the user of such a stream to Read all the data that's written.